File: verman-py/project_manager.py
"""
Project lifecycle management for VerMan.
"""

import os
import shutil
from datetime import datetime
from typing import Optional

from database import DatabaseManager
from file_manager import FileManager
from logger import logger as app_logger


class ProjectManager:
    """Manage create/open/close/delete operations for a workspace."""

    def __init__(self):
        self.current_project_path: Optional[str] = None
        self.db_manager: Optional[DatabaseManager] = None
        self.file_manager: Optional[FileManager] = None

    def create_project(self, workspace_path: str) -> bool:
        try:
            workspace_path = os.path.abspath(workspace_path)
            if not os.path.isdir(workspace_path):
                return False

            db_path = os.path.join(workspace_path, ".verman.db")
            if os.path.exists(db_path):
                return False

            self._create_ignore_file(workspace_path)
            self.db_manager = DatabaseManager(db_path)
            self.db_manager.set_config("project_path", workspace_path)
            self.db_manager.set_config("create_time", self._get_current_time())

            self.file_manager = FileManager(workspace_path)
            self.current_project_path = workspace_path
            return True
        except Exception as exc:
            app_logger.error(f"创建项目失败: {exc}")
            return False

    def open_project(self, workspace_path: str) -> bool:
        try:
            workspace_path = os.path.abspath(workspace_path)
            if not os.path.isdir(workspace_path):
                return False

            db_path = os.path.join(workspace_path, ".verman.db")
            if not os.path.exists(db_path):
                return False

            if DatabaseManager.requires_migration(db_path):
                self._backup_database(db_path)

            self.db_manager = DatabaseManager(db_path)
            stored_path = self.db_manager.get_config("project_path")
            if stored_path and stored_path != workspace_path:
                app_logger.warning(
                    f"存储的项目路径({stored_path})与当前路径({workspace_path})不匹配"
                )

            self.file_manager = FileManager(workspace_path)
            self.current_project_path = workspace_path
            return True
        except Exception as exc:
            app_logger.error(f"打开项目失败: {exc}")
            return False

    def close_project(self):
        if self.db_manager:
            self.db_manager.close()
            self.db_manager = None

        self.file_manager = None
        self.current_project_path = None

    def delete_project(self, workspace_path: str) -> bool:
        try:
            workspace_path = os.path.abspath(workspace_path)
            db_path = os.path.join(workspace_path, ".verman.db")
            if not os.path.exists(db_path):
                return False

            if self.current_project_path == workspace_path:
                self.close_project()

            os.remove(db_path)
            return True
        except Exception as exc:
            app_logger.error(f"删除项目失败: {exc}")
            return False

    def is_project_open(self) -> bool:
        return self.current_project_path is not None and self.db_manager is not None

    def get_current_project_path(self) -> Optional[str]:
        return self.current_project_path

    def get_database_manager(self) -> Optional[DatabaseManager]:
        return self.db_manager

    def get_file_manager(self) -> Optional[FileManager]:
        return self.file_manager

    def get_project_info(self) -> Optional[dict]:
        if not self.is_project_open():
            return None

        try:
            info = {
                "project_path": self.db_manager.get_config("project_path"),
                "create_time": self.db_manager.get_config("create_time"),
            }

            versions = self.db_manager.get_all_versions()
            info["version_count"] = len(versions)
            if versions:
                info["latest_version"] = versions[0]["version_number"]
                info["latest_time"] = versions[0]["create_time"]
            else:
                info["latest_version"] = "无"
                info["latest_time"] = "无"
            return info
        except Exception as exc:
            app_logger.error(f"获取项目信息失败: {exc}")
            return None

    def _backup_database(self, db_path: str):
        timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
        backup_path = f"{db_path}.bak.{timestamp}"
        shutil.copy2(db_path, backup_path)
        app_logger.info(f"已备份旧数据库: {backup_path}")

    def _create_ignore_file(self, workspace_path: str):
        ignore_file_path = os.path.join(workspace_path, ".vermanignore")
        if os.path.exists(ignore_file_path):
            return

        ignore_content = """# VerMan 忽略文件
# 此文件用于指定版本管理中需要忽略的文件和目录

# 版本管理数据库和备份
.verman.db
.verman.db-shm
.verman.db-wal
.verman.db-journal
.verman.db.bak.*
.verman_backup/

# Python 相关
__pycache__/
*.py[cod]
*$py.class
*.so
.Python
build/
dist/
*.egg-info/
*.egg

# 虚拟环境
.env
.venv/
env/
venv/
ENV/

# IDE 相关
.vscode/
.idea/
*.swp
*.swo
*~

# 系统文件
.DS_Store
Thumbs.db
desktop.ini

# 临时文件
*.tmp
*.temp
*.log
*.bak
*.backup
"""

        with open(ignore_file_path, "w", encoding="utf-8") as file_handle:
            file_handle.write(ignore_content)

    def _get_current_time(self) -> str:
        return datetime.now().strftime("%Y-%m-%d %H:%M:%S")

    def __del__(self):
        self.close_project()

File: verman-rust/src-tauri/src/project_manager.rs
use std::path::{Path, PathBuf};

use crate::database::DatabaseManager;
use crate::file_manager::FileManager;
use crate::file_watcher::FileWatcher;
use crate::project_paths;
use crate::version_manager::VersionManager;

pub struct ProjectManager {
    pub current_project_path: Option<PathBuf>,
    pub version_manager: Option<VersionManager>,
    pub file_watcher: Option<FileWatcher>,
}

impl ProjectManager {
    pub fn new() -> Self {
        Self {
            current_project_path: None,
            version_manager: None,
            file_watcher: None,
        }
    }

    pub fn create_project(
        &mut self,
        workspace_path: &Path,
        config_ignore_patterns: Vec<String>,
    ) -> Result<bool, String> {
        let workspace_path = workspace_path.to_owned();
        if !workspace_path.is_dir() {
            return Ok(false);
        }
        if project_paths::is_project_workspace(&workspace_path) {
            return Ok(false);
        }

        project_paths::ensure_metadata_dir(&workspace_path)
            .map_err(|e| format!("Failed to create metadata dir: {}", e))?;
        self.create_ignore_file(&workspace_path);

        let db_path = project_paths::get_project_database_path(&workspace_path);
        let db_manager =
            DatabaseManager::new(&db_path).map_err(|e| format!("Failed to init DB: {}", e))?;
        db_manager
            .set_config("project_path", &workspace_path.to_string_lossy())
            .map_err(|e| format!("Failed to set config: {}", e))?;
        db_manager
            .set_config(
                "create_time",
                &chrono::Local::now().format("%Y-%m-%d %H:%M:%S").to_string(),
            )
            .map_err(|e| format!("Failed to set config: {}", e))?;

        let file_manager = FileManager::new(&workspace_path);
        let vm = VersionManager::new(db_manager, file_manager, config_ignore_patterns);

        self.current_project_path = Some(workspace_path);
        self.version_manager = Some(vm);
        Ok(true)
    }

    pub fn open_project(
        &mut self,
        workspace_path: &Path,
        config_ignore_patterns: Vec<String>,
    ) -> Result<bool, String> {
        let workspace_path = workspace_path.to_owned();
        if !workspace_path.is_dir() {
            return Ok(false);
        }

        self.migrate_legacy_layout(&workspace_path);

        let db_path = match project_paths::find_existing_database_path(&workspace_path) {
            Some(p) => p,
            None => return Ok(false),
        };

        let db_manager =
            DatabaseManager::new(&db_path).map_err(|e| format!("Failed to open DB: {}", e))?;

        let file_manager = FileManager::new(&workspace_path);
        let vm = VersionManager::new(db_manager, file_manager, config_ignore_patterns);

        self.current_project_path = Some(workspace_path);
        self.version_manager = Some(vm);
        Ok(true)
    }

    pub fn close(&mut self) {
        if let Some(ref vm) = self.version_manager {
            if let Ok(fm) = vm.file_manager.lock() {
                fm.save_hash_cache();
            }
        }
        self.file_watcher = None;
        self.version_manager = None;
        self.current_project_path = None;
    }

    pub fn start_watcher(&mut self, path: &Path, app_handle: tauri::AppHandle) {
        self.file_watcher = FileWatcher::start(path.to_path_buf(), app_handle).ok();
    }

    pub fn is_open(&self) -> bool {
        self.current_project_path.is_some() && self.version_manager.is_some()
    }

    pub fn get_project_path(&self) -> Option<&Path> {
        self.current_project_path.as_deref()
    }

    fn migrate_legacy_layout(&self, workspace_path: &Path) {
        let legacy_db = project_paths::get_legacy_database_path(workspace_path);
        let new_db = project_paths::get_project_database_path(workspace_path);
        if !legacy_db.exists() || new_db.exists() {
            return;
        }

        if let Err(e) = project_paths::ensure_metadata_dir(workspace_path) {
            tracing::warn!(error = %e, "Cannot create metadata dir for migration");
            return;
        }

        // Migrate database sidecar files
        for (legacy, new) in project_paths::iter_database_sidecar_paths(&legacy_db)
            .into_iter()
            .zip(project_paths::iter_database_sidecar_paths(&new_db))
        {
            if legacy.exists() {
                if let Err(e) = std::fs::rename(&legacy, &new) {
                    tracing::warn!(error = %e, "Failed to migrate {:?}", legacy);
                } else {
                    tracing::info!("Migrated {:?} -> {:?}", legacy, new);
                }
            }
        }

        // Migrate backup files
        let pattern = format!("{}.bak.*", legacy_db.display());
        if let Ok(entries) = glob::glob(&pattern) {
            for entry in entries.flatten() {
                let suffix = entry
                    .to_string_lossy()
                    .split(&format!("{}.bak.", legacy_db.display()))
                    .nth(1)
                    .unwrap_or("")
                    .to_string();
                let new_backup = format!("{}.bak.{}", new_db.display(), suffix);
                if let Err(e) = std::fs::rename(&entry, Path::new(&new_backup)) {
                    tracing::warn!(error = %e, "Failed to migrate backup {:?}", entry);
                }
            }
        }

        // Migrate legacy backup directory
        let legacy_backup_dir = project_paths::get_legacy_backup_dir(workspace_path);
        let backup_dir = project_paths::get_backup_dir(workspace_path);
        if legacy_backup_dir.is_dir() {
            if backup_dir.exists() {
                Self::merge_directories(&legacy_backup_dir, &backup_dir);
                std::fs::remove_dir_all(&legacy_backup_dir).ok();
            } else {
                std::fs::rename(&legacy_backup_dir, &backup_dir).ok();
            }
            tracing::info!("Migrated backup dir {:?} -> {:?}", legacy_backup_dir, backup_dir);
        }
    }

    fn merge_directories(source: &Path, target: &Path) {
        let walker = walkdir::WalkDir::new(source);
        for entry in walker.into_iter().flatten() {
            if entry.file_type().is_file() {
                let relative = entry.path().strip_prefix(source).unwrap_or(entry.path());
                let dest = target.join(relative);
                if let Some(parent) = dest.parent() {
                    std::fs::create_dir_all(parent).ok();
                }
                std::fs::rename(entry.path(), &dest).ok();
            }
        }
    }

    fn create_ignore_file(&self, workspace_path: &Path) {
        let ignore_path = project_paths::get_ignore_file_path(workspace_path);
        if ignore_path.exists() {
            return;
        }

        let content = r#"# VerMan 忽略文件
# 该文件用于指定版本管理中需要忽略的文件和目录

# VerMan 元数据
.verman/
.verman.db
.verman.db-shm
.verman.db-wal
.verman.db-journal
.verman.db.bak.*
.verman_backup/

# Python 相关
__pycache__/
*.py[cod]
*$py.class
*.so
.Python
build/
dist/
*.egg-info/
*.egg

# 虚拟环境
.env
.venv/
env/
venv/
ENV/

# IDE 相关
.vscode/
.idea/
*.swp
*.swo
*~

# 系统文件
.DS_Store
Thumbs.db
desktop.ini

# 临时文件
*.tmp
*.temp
*.log
*.bak
*.backup
"#;

        std::fs::write(&ignore_path, content).ok();
    }
}
